core/trace/src/main/scala/org/typelevel/otel4s/trace/SpanContext.scala

```diff
@@ -104,14 +122,147 @@ object SpanContext {
+    *   the trace state of the span context
+    *
+    * @param remote
+    *   whether the span is propagated from the remote parent or not
+    */
+  def apply(
+      traceId: ByteVector,
+      spanId: ByteVector,
+      traceFlags: TraceFlags,
+      traceState: TraceState,
+      remote: Boolean
+  ): SpanContext =
+    if (TraceId.isValid(traceId) && SpanId.isValid(spanId)) {
+      createInternal(
+        traceId = traceId,
+        spanId = spanId,
+        traceFlags = traceFlags,
+        traceState = traceState,
+        remote = remote,
+        isValid = true
+      )
+    } else {
+      createInternal(
+        traceId = TraceId.Invalid,
+        spanId = SpanId.Invalid,
+        traceFlags = traceFlags,
+        traceState = traceState,
+        remote = remote,
+        isValid = false
+      )
     }
+
+  implicit val spanContextHash: Hash[SpanContext] =
+    Hash.by { ctx =>
+      (
+        ctx.traceIdHex,
+        ctx.spanIdHex,
+        ctx.traceFlags,
+        ctx.traceState,
+        ctx.isValid,
+        ctx.isRemote
+      )
+    }
+
+  implicit val spanContextShow: Show[SpanContext] =
+    Show.show { ctx =>
+      show"SpanContext{" +
+        show"traceId=${ctx.traceIdHex}, " +
+        show"spanId=${ctx.spanIdHex}, " +
+        show"traceFlags=${ctx.traceFlags}, " +
+        show"traceState=${ctx.traceState}, " +
+        show"remote=${ctx.isRemote}, " +
+        show"valid=${ctx.isValid}}"
+    }
+
+  /** Creates a new [[SpanContext]] with the given identifiers and options.
+    *
+    * If the id validation isn't skipped and the `traceId` or the `spanId` are
+    * invalid (ie. do not conform to the requirements for hexadecimal ids of the
+    * appropriate lengths), both will be replaced with the standard "invalid"
+    * versions (i.e. all '0's).
+    *
+    * '''Note''': the method is for the internal use only. It is not supposed to
+    * be publicly available.
+    *
+    * @see
+    *   [[apply]]
+    *
+    * @param traceId
+    *   the trace identifier of the span context
+    *
+    * @param spanId
+    *   the span identifier of the span context
+    *
+    * @param traceFlags
+    *   the trace flags of the span context
+    *
+    * @param traceState
+    *   the trace state of the span context
+    *
+    * @param remote
+    *   whether the span is propagated from the remote parent or not
+    *
+    * @param isValid
+    *   whether the span is valid or not
+    */
+  private[otel4s] def createInternal(
+      traceId: ByteVector,
+      spanId: ByteVector,
+      traceFlags: TraceFlags,
+      traceState: TraceState,
+      remote: Boolean,
+      isValid: Boolean
+  ): SpanContext =
+    SpanContextImpl(
+      traceId = traceId,
+      traceIdHex = traceId.toHex,
+      spanId = spanId,
+      spanIdHex = spanId.toHex,
+      traceFlags = traceFlags,
+      traceState = traceState,
+      isRemote = remote,
+      isValid = isValid
+    )
+
+  private final case class SpanContextImpl(
```

Comment:
extreme nitpick: should this just be called `Impl`, as it's already inside the `SpanContext` companion?